xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1125 @@
+		ConnectivityState: connectivity.Ready,
+		Picker:            &neverErrPicker{},
+	})
+
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for a part UpdateState call on the ClientConn")
+	case state := <-tcc.NewStateCh:
+		if state != connectivity.Ready {
+			t.Fatal("wanted connectivity state ready")
+		}
+	}
+
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for a part UpdateState call on the ClientConn")
+	case picker := <-tcc.NewPickerCh:
+		// Should receive a never err picker.
+		_, err := picker.Pick(balancer.PickInfo{})
+		if err != nil {
+			t.Fatalf("Graceful Switch Balancer should have sent a never error picker, which is a picker that never errors, instead received err %v", err)
+		}
+	}
+
+	// An explicit call to switchTo, even if the same type, should cause the
+	// balancer to build a new balancer for pending.
+	gsb.SwitchTo(mockBalancerBuilder1{})
+	if gsb.balancerPending == nil {
+		t.Fatal("balancerPending was not built out when another SwitchTo() call should've triggered the pending balancer to build")
+	}
+
+	// A Client Conn update received should be forwarded to the new pending LB
+	// policy, and not the current one.
+	gsb.UpdateClientConnState(ccs)
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	defer cancel()
+	if err := gsb.balancerCurrent.Balancer.(*mockBalancer).waitForClientConnUpdate(ctx, ccs); err == nil {
+		t.Fatal("balancerCurrent should not have received a client Conn update if there is a pending LB policy")
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := gsb.balancerPending.Balancer.(*mockBalancer).waitForClientConnUpdate(ctx, ccs); err != nil {
+		t.Fatalf("error in ClientConnState update: %v", err)
+	}
+
+	// If the pending LB reports that is CONNECTING, no update should be sent to
+	// the Client Conn.
+	gsb.balancerPending.Balancer.(*mockBalancer).updateState(balancer.State{
+		ConnectivityState: connectivity.Connecting,
+	})
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	defer cancel()
+	select {
+	case <-tcc.NewStateCh:
+		t.Fatal("Pending LB reporting CONNECTING should not forward up to the ClientConn")
+	case <-ctx.Done():
+	}
+
+	currBal := gsb.balancerCurrent.Balancer.(*mockBalancer)
+	// If the pending LB reports a state other than CONNECTING, the pending LB
+	// is logically warmed up, and the ClientConn should be updated with the
+	// State and Picker to start using the new policy. The pending LB policy
+	// should also be switched into the current LB.
+	gsb.balancerPending.Balancer.(*mockBalancer).updateState(balancer.State{
+		ConnectivityState: connectivity.Ready,
+		Picker:            &neverErrPicker{},
+	})
+
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
```

Comment:
Please fix the rest of this file with comments from above. Will take another look after that.

Reply:
Done.